### ISSUE TITLE
HOL_ERR to pick up term and type arguments

### ISSUE THREAD

**Issue by mn200** [HOL-Theorem-Prover/HOL#622]
Change the fundamental exception to give exception-raisers the option of attaching relevant terms and types.  For the sake of backwards compatibility, we can redefine helpers like `mk_HOL_ERR` and friends (creating values with empty lists).  My suggestion would be something like
```
exception HOL_ERR of {origin_structure : string, origin_function : string, message : string,
                                        terms : term list, types : hol_type list}
```

**Comment by barakeel** [comment]
Most of the exception won't use these fields. 
So there might be a negligible amount of a (memory/cpu) waste.
Could you provide motivating examples that require these changes?

**Comment by konrad-slind** [comment]
I agree with Thibault. Some additional ..._ERR exceptions could be perhaps
be declared for more demanding situations, but HOL_ERR works OK the
majority of the time.

**Comment by mn200** [comment]
I was actually thinking that it might be worthwhile to further *add* a location field.  The parser could use these additional fields almost everywhere.  Tactics that fail because a term has the wrong sort of shape would also use them. Decision procedures likewise. 

If we were storing large numbers of exceptions, I might be worried about memory wastage, but almost all exceptions will live only transiently such that I'd be surprised if there were ever more than single digit numbers of them ever allocated at once.

**Comment by mn200** [comment]
For example, try 
```
   grep -r --include='*.sml' failwith.*term_to_string src
```
or
```
   grep -r --include='*.sml' ERR.*term_to_string src
```

This will a subset of the occurrences because it requires the `term_to_string` to appear on the same line as the `ERR` or `failwith`, but it still gets a number of hits.  These are much worse for efficiency than simply being able to return the term in the proposed list field.  Pretty-printing is expensive (*i.e.,* slow) and the generated string is *less* helpful than actually being able to get your hands on the SML value.

**Comment by jeremydaw** [comment]
Rather than trying to add additional fields of whatever type anyone will ever find useful, how about something like the following:

exception HOL_ERR of {origin_structure : string, origin_function : string,
  message : string, terms : term list, types : hol_type list,
  further_info : exn list}
  
exception TT of term * hol_type ;

val tt = (``0``, ``:num``) ;

HOL_ERR {origin_structure = "os", origin_function = "of",
  message = "msg", terms = [], types = [], further_info = [TT tt]} ;

**Comment by mn200** [comment]
This is a good idea, though I'd perhaps prefer to use the slight abstraction provided by our existing `UniversalType` to get the same effect.  I'd keep the common cases of terms and types, and maybe also the location information, but there's little harm in also having the “extra info” field as you describe.

**Comment by jeremydaw** [comment]
Yes, I hadn't known about ```UniversalType```.

As with the exception constructors in my suggestion, users (ie those who create exceptions using this feature) would need to make sure the destructor function is available (signature-wise) to the end user (ie the one who sees the exception raised)

**Comment by mn200** [comment]
Absolutely!

**Comment by barakeel** [comment]
Indeed term_to_string is slow. I understand the reason now.

**Comment by talsewell** [comment]
Another option is to use lazy strings rather than terms, types, etc, since the
most common case is for them to be printed to strings.

I wrote that up in #928  which I now realise is probably just a part of this.

**Comment by mn200** [comment]
One issue with this is that the `HOL_ERR` exception declaration has to happen in a context where terms and types are in scope.  Currently, it’s declared *before* the kernel's types are established, which means that kernel operations and even functions in `Lib` can return it. 

Though declaring the exception before term functions were defined, but after the type was set up, would allow term functions to throw it, type functions couldn't.  Moreover, my experience elsewhere suggests that the terms put into such an exception wouldn't get pretty-printed well.  The `UniversalType` argument might be a better approach.

**Comment by ordinarymath** [comment]
As I've noted on the zulip. mosml currently does not support printing an exception inside of a exception. The `UniversalType` in mosml implemented using exceptions
